src/analysis/special_functions/trigonometric.lean

```diff
@@ -1478,88 +1478,332 @@ end
+eq_comm.trans arcsin_eq_zero_iff
+
+@[simp] lemma arcsin_pos {x : ℝ} : 0 < arcsin x ↔ 0 < x :=
+lt_iff_lt_of_le_iff_le arcsin_nonpos
+
+@[simp] lemma arcsin_lt_zero {x : ℝ} : arcsin x < 0 ↔ x < 0 :=
+lt_iff_lt_of_le_iff_le arcsin_nonneg
+
+@[simp] lemma arcsin_lt_pi_div_two {x : ℝ} : arcsin x < π / 2 ↔ x < 1 :=
+(arcsin_lt_iff_lt_sin' (right_mem_Ioc.2 $ neg_lt_self pi_div_two_pos)).trans $
+  by rw sin_pi_div_two
+
+@[simp] lemma neg_pi_div_two_lt_arcsin {x : ℝ} : -(π / 2) < arcsin x ↔ -1 < x :=
+(lt_arcsin_iff_sin_lt' $ left_mem_Ico.2 $ neg_lt_self pi_div_two_pos).trans $
+  by rw [sin_neg, sin_pi_div_two]
+
+@[simp] lemma arcsin_eq_pi_div_two {x : ℝ} : arcsin x = π / 2 ↔ 1 ≤ x :=
+⟨λ h, not_lt.1 $ λ h', (arcsin_lt_pi_div_two.2 h').ne h, arcsin_of_one_le⟩
+
+@[simp] lemma pi_div_two_eq_arcsin {x} : π / 2 = arcsin x ↔ 1 ≤ x :=
+eq_comm.trans arcsin_eq_pi_div_two
+
+@[simp] lemma pi_div_two_le_arcsin {x} : π / 2 ≤ arcsin x ↔ 1 ≤ x :=
+(arcsin_le_pi_div_two x).le_iff_eq.trans pi_div_two_eq_arcsin
+
+@[simp] lemma arcsin_eq_neg_pi_div_two {x : ℝ} : arcsin x = -(π / 2) ↔ x ≤ -1 :=
+⟨λ h, not_lt.1 $ λ h', (neg_pi_div_two_lt_arcsin.2 h').ne' h, arcsin_of_le_neg_one⟩
+
+@[simp] lemma neg_pi_div_two_eq_arcsin {x} : -(π / 2) = arcsin x ↔ x ≤ -1 :=
+eq_comm.trans arcsin_eq_neg_pi_div_two
+
+@[simp] lemma arcsin_le_neg_pi_div_two {x} : arcsin x ≤ -(π / 2) ↔ x ≤ -1 :=
+(neg_pi_div_two_le_arcsin x).le_iff_eq.trans arcsin_eq_neg_pi_div_two
+
+lemma maps_to_sin_Ioo : maps_to sin (Ioo (-(π / 2)) (π / 2)) (Ioo (-1) 1) :=
+λ x h, by rwa [mem_Ioo, ← arcsin_lt_pi_div_two, ← neg_pi_div_two_lt_arcsin,
+  arcsin_sin h.1.le h.2.le]
+
+/-- `real.sin` as a `local_homeomorph` between `(-π / 2, π / 2)` and `(-1, 1)`. -/
+@[simp] def sin_local_homeomorph : local_homeomorph ℝ ℝ :=
+{ to_fun := sin,
+  inv_fun := arcsin,
+  source := Ioo (-(π / 2)) (π / 2),
+  target := Ioo (-1) 1,
+  map_source' := maps_to_sin_Ioo,
+  map_target' := λ y hy, ⟨neg_pi_div_two_lt_arcsin.2 hy.1, arcsin_lt_pi_div_two.2 hy.2⟩,
+  left_inv' := λ x hx, arcsin_sin hx.1.le hx.2.le,
+  right_inv' := λ y hy, sin_arcsin hy.1.le hy.2.le,
+  open_source := is_open_Ioo,
+  open_target := is_open_Ioo,
+  continuous_to_fun := continuous_sin.continuous_on,
+  continuous_inv_fun := continuous_arcsin.continuous_on }
+
+lemma cos_arcsin_nonneg (x : ℝ) : 0 ≤ cos (arcsin x) :=
+cos_nonneg_of_mem_Icc ⟨neg_pi_div_two_le_arcsin _, arcsin_le_pi_div_two _⟩
+
+lemma cos_arcsin {x : ℝ} (hx₁ : -1 ≤ x) (hx₂ : x ≤ 1) : cos (arcsin x) = sqrt (1 - x ^ 2) :=
+have sin (arcsin x) ^ 2 + cos (arcsin x) ^ 2 = 1 := sin_sq_add_cos_sq (arcsin x),
+begin
+  rw [← eq_sub_iff_add_eq', ← sqrt_inj (pow_two_nonneg _) (sub_nonneg.2 (sin_sq_le_one (arcsin x))),
+    pow_two, sqrt_mul_self (cos_arcsin_nonneg _)] at this,
+  rw [this, sin_arcsin hx₁ hx₂],
+end
+
+lemma deriv_arcsin_aux {x : ℝ} (h₁ : x ≠ -1) (h₂ : x ≠ 1) (n : with_top ℕ) :
+  has_deriv_at arcsin (1 / sqrt (1 - x ^ 2)) x ∧ times_cont_diff_at ℝ n arcsin x :=
```

Comment:
It doesn't really matter, but I think the statement is nicer if the `(n : with_top ℕ)` before the colon becomes a `∧ ∀ (n : with_top ℕ), ...` after the colon, to avoid the unnecessary argument to the left part of the conjunction.